File: programmers/kakao/level3/표현_가능한_이진트리.py
from math import log2

def solution(numbers):
    answer = []
    
    def dfs(x, parent):
        if parent == "0":
            if not all(child == "0" for child in x):
                return False
        if len(x) == 1:
            return True
        mid = len(x)>>1
        return dfs(x[:mid], x[mid]) and dfs(x[mid+1:], x[mid])
                
    for num in numbers:
        num2 = bin(num)[2:]
        h = int(log2(len(num2)))+1
        L = pow(2, h)-1
        num2 = "0"*(L-len(num2)) + num2
        answer.append(1 if dfs(num2, num2[L//2]) else 0)

    return answer


numbers = [7, 42, 5]
print(solution(numbers)) # [1, 1, 0]

numbers = [63, 111, 85]
print(solution(numbers)) # [1, 1, 0]

File: Java/programmers/kakao/level3/표현가능한이진트리.java
package CodeTest.Java.programmers.kakao.level3;

public class 표현가능한이진트리 {
    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];

        for (int i=0; i<numbers.length; i++) {
            String st = Long.toBinaryString(numbers[i]);
            int j = 0;
            while ((int)Math.pow(2,j)-1 < st.length()) j++;
            while ((int)Math.pow(2,j)-1 != st.length()) st = "0"+st;
            answer[i] = dfs(st) ? 1 : 0;
        }

        return answer;
    }

    private static boolean dfs(String x) {
        int mid = (x.length()-1)/2;
        char root = x.charAt(mid);
        String left = x.substring(0, mid);
        String right = x.substring(mid+1);

        int l = left.length(), r = right.length();
        if (root=='0' && (left.charAt((l-1)/2)=='1' || right.charAt((r-1)/2)=='1')) {
            return false;
        }

        if (l >= 3) return dfs(left) && dfs(right);
        
        return true;
    }
}
